nemo_skills/inference/generate.py

```diff
@@ -257,48 +258,91 @@ def async_loop(cfg, data, llm, prompt, extra_stop_phrases, extra_generate_params
     if len(data) == 0:  # we might not have any examples if skip_filled=True
         return
 
-    LOG.warning("Async loop is submitting all data for inference - batch_size parameter is ignored!")
+    LOG.warning("Maintaining 1000 concurrent requests throughout execution.")
 
-    # submitting all data at ones
-    generation_ids = llm.generate_async(
-        prompts=[prompt.fill(dp) for dp in data],
-        stop_phrases=combine_stop_phrases(prompt.stop_phrases, extra_stop_phrases),
-        **asdict(cfg.inference),
-        **extra_generate_params,
-    )
+    
+    
+    request_queue = list(range(len(data)))  # Queue of unsubmitted task indices
+    in_progress = {}  # Track ongoing requests {index: generation_id}
+    
 
-    # setting buffering=1 to force to dump the output after every line, so that we can see intermediate generations
-    with open(cfg.output_file + '-async', "at" if cfg.skip_filled else "wt", encoding="utf-8", buffering=1) as fout:
-        pbar = tqdm(total=len(generation_ids), desc="Remaining generations")
-        prev_remaining = len(generation_ids)
-        while True:
-            remaining_ids = [generation_id for generation_id in generation_ids if generation_id is not None]
-            curr_remaining = len(remaining_ids)
-            if curr_remaining == 0:
-                break
-            # Update progress bar with completed generations
-            if curr_remaining < prev_remaining:
-                pbar.update(prev_remaining - curr_remaining)
-                prev_remaining = curr_remaining
-
-            remaining_positions = [
-                idx for idx, generation_id in enumerate(generation_ids) if generation_id is not None
-            ]
-            generations = llm.get_generations(remaining_ids)
-            for gen_pos, gen_dict in zip(remaining_positions, generations):
-                if gen_dict['generation'] is not None:  # will be None until done
-                    generation_ids[gen_pos] = None
-                    gen_dict[cfg.generation_key] = gen_dict.pop("generation")
-                    for key in gen_dict:
-                        data[gen_pos].pop(key, None)
-                    gen_dict.update(data[gen_pos])
 
-                    # insert async position information
-                    gen_dict[cfg.async_position_key] = original_positions[gen_pos]
+    with open(cfg.output_file + "-async", "at" if cfg.skip_filled else "wt", encoding="utf-8", buffering=1) as fout:
+        pbar = tqdm(total=len(data), desc="Processing requests")
 
-                    fout.write(json.dumps(gen_dict) + "\n")
+        # **Step 1: Pre-fill `in_progress` with the first 2000 tasks**
+        while len(in_progress) < cfg.max_concurrent_requests and request_queue:
+            idx = request_queue.pop(0)
```

Comment:
if you want to use this .pop / .append interface, better to switch to a real dequeue from python as popping a list is very inefficient